frb_dart/lib/src/ffi/io.dart

```diff
@@ -51,26 +37,134 @@ class DartApiDl {
 /// This class is only for internal usage.
 class WireSyncReturnStruct extends ffi.Struct {
   /// Not to be used by normal users, but has to be public for generated code
-  external ffi.Pointer<ffi.Uint8> ptr;
+  external DartCObject data;
 
-  /// Not to be used by normal users, but has to be public for generated code
-  @ffi.Int32()
-  external int len;
+  @ffi.Bool()
+  external bool isSuccess;
 
-  /// Not to be used by normal users, but has to be public for generated code
-  @ffi.Uint8()
-  external int success;
+  external ffi.Pointer<ffi.Void> ptr;
+
+  dynamic intoDart() => data.intoDart();
+}
 
-  Uint8List? get buffer {
-    if (ptr.address == 0) {
-      return null;
+extension DartCObjectWireSyncReturn on DartCObject {
+  dynamic intoDart() {
+    switch (ty) {
+      case 0: // DartNull
+        return null;
+      case 1: // DartBool
+        return value.as_bool;
+      case 2: // DartInt32
+        return value.as_int32;
+      case 3: // DartInt64
+        return value.as_int64;
+      case 4: // DartDouble
+        return value.as_double;
+
+      case 5: // DartString
+        int len = 0;
+        while (value.as_string.elementAt(len).value != 0) {
+          len++;
+        }
+        return utf8.decode(value.as_string.cast<ffi.Uint8>().asTypedList(len));
```

Comment:
nit: Are we sure the string is `\0`-terminated? 
If so, we should add a test about it to ensure. For example, deliberately create a Rust `String` with `\0` *in the middle* and assert results in the dart side.
In addition, maybe add a line of comment pointing to where the "0-terminated" is specified
If not,  we may need to change the way

Reply:
[We are sure yes](https://github.com/sunshine-protocol/allo-isolate/blob/71b9760993d64ef46794176ca276d1cc637b2599/src/into_dart.rs#L106)! Will add the comment. [std::ffi::CString](https://doc.rust-lang.org/nightly/std/ffi/struct.CString.html) asserts this guarantee.

Because of `allo_isolate`'s implementation, a string containing the _nul_ character will be unwrapped as default (empty). Is this really our job to test this?

Reply:
LGTM. Maybe just add a very simple test: `api.rs` has sth like `fn f() -> String { String::new("abc\0def") }` and `main.dart` has `expect(f(), "")`. Just a few lines and we are more sure;)

Reply:
Added the test for async as well

Reply:
These new tests failed for web as this nul-character behavior is specific to native env. In wasm, the string returned still contains the nul-character.

This should be the ideal scenario for native as well but it's out of the scope of this PR. Just reproducing the existing async behavior.

If it's believed to be important, an issue could be created!

Reply:
> These new tests failed for web as this nul-character behavior is specific to native env. In wasm, the string returned still contains the nul-character.

Anything reasonable is accpetable, as long as we do not see crashes, memory leaks, invalid memory access, etc.